src/ray/util/shared_lru.h

```diff
@@ -0,0 +1,135 @@
+// SharedLruCache is a LRU cache, with all entries shared, which means a single entry
+// could be accessed by multiple getters. When `Get`, a copy of the value is returned, so
+// for heavy-loaded types it's suggested to wrap with `std::shared_ptr<>`.
+//
+// Example usage:
+// SharedLruCache<std::string, std::string> cache{cap};
+// // Put a key-value pair into cache.
+// cache.Put("key", "val");
+//
+// // Get a key-value pair from cache.
+// auto val = ache.Get("key");
+// // Check and consume `val`.
+//
+// TODO(hjiang):
+// 1. Add template arguments for key hash and key equal, to pass into absl::flat_hash_map.
+// 2. Provide a key hash wrapper to save a copy.
+// 3. flat hash map supports heterogeneous lookup, expose `KeyLike` templated interface.
+
+#pragma once
+
+#include <cstdint>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <string>
+#include <utility>
+
+#include "absl/container/flat_hash_map.h"
+
+namespace ray::utils::container {
+
+template <typename Key, typename Val>
+class SharedLruCache final {
+ public:
+  using key_type = Key;
+  using mapped_type = Val;
+
+  // A `max_entries` of 0 means that there is no limit on the number of entries
+  // in the cache.
+  explicit SharedLruCache(size_t max_entries) : max_entries_(max_entries) {}
```

Comment:
reserve cache_ with max_entries

Reply:
No it's intentional. For example, if the `max_entries` are super large, you don't want to allocate big chunk of memory right now.

Reply:
A better approach might be have an extra argument as part of the ctor. But not sure if it's really needed.

Reply:
I wrote the similarly same library in my prev company, my observation is customers usually set super large `capacity_`, which they never need.

Reply:
makes sense 👍🏽